### ISSUE TITLE
BUG: If validation errors occur while entering a Petition and Clerk attempts to enter the data in a field, the screen jumps.

### ISSUE THREAD

**Issue by ttlenard** [ustaxcourt/ef-cms#8468]
**Describe the Bug**
When a Petitions clerk is entering in a paper petition and they click Save before all of the required fields are entered, they will receive validation errors indicating that they need to enter in the required data.  When this happens, the clerk would then find the field that they missed filling out initially and click their cursor in that field. When they do this, the screen jumps up to the top, back to the validation error message.  If they scroll down, they can continue to enter their data, but if there is another field that they initially missed filling out, then when they put their cursor in that field, and attempt to type, the screen will jump again.  I also noticed that when I clicked anywhere on the screen after receiving a validation error message, the screen would also jump to the top.

Question: Is this similar to bug ustaxcourt/ef-cms#8531?

**Business Impact/Reason for Severity**
Low

**In which environment did you see this bug?**
Test

**Who were you logged in as?**
Petitions Clerk

**What were you doing when you discovered this bug? (Using the application, demoing, smoke tests, testing other functionality, etc.)**
Testing other functionality

**To Reproduce**
1. Log in as a Petitions Clerk
2. Click on Document QC
3. Click on Start a Petition
4. Begin to fill out the Petition fields, but leave some of the required fields blank
5. Click Create Case
6. You will receive a validation error, listing the fields that are required
7. Scroll down to the fields that you skipped 
8. Add your cursor to the field (note: not ALL of the fields jump. Mailing address didn't, but city, state, zip, phone do on the Petitioner tab.)
9. The screen jumps to the top of the screen to the validation message again
10. You have to scroll down and find the field again to enter the data.
11. The jumpy screen also occurs when you just click on the screen after the validation error is thrown.

**Expected Behavior**
Screen shouldn't jump around after a validation error.  I think it is great on the initial error, but is annoying when it keeps jumping the screen back to the top when you are attempting to add your data to the required fields that you missed initially.

**Actual Behavior**
A clear and concise description of what actually happened.
Screen continues to jump up to the validation message when you click on the field that you initially missed. You have to scroll back down the find the field that you missed and THEN you can add data.

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Cause of Bug, If Known**

**Process for Logging a Bug:**
* Complete the above information
* Add a severity tag (Critical, High Severity, Medium Severity or Low Severity). See below for priority definition. 

**Severity Definition:**
* Critical Defect
Blocks entire system's or module’s functionality
No workarounds available
Testing cannot proceed further without bug being fixed.

* High-severity Defect
Affects key functionality of an application
There's a workaround, but not obvious or easy
App behaves in a way that is strongly different from the one stated in the requirements

* Medium-severity Defect
A minor function does not behave in a way stated in the requirements.
Workaround is available and easy

* Low-severity Defect
Mostly related to an application’s UI
Doesn't need a workaround, because it doesn't impact functionality

## Definition of Ready for Bugs(Created 10-4-21)
Definition used: A failure or flaw in the system which produces an incorrect or undesired result that deviates from the expected result or behavior. (Note: Expected results are use cases that have been documented in past user stories as acceptance criteria and test cases, and do not include strange behavior unrelated to use cases.)

The following criteria must be met in order for the development team to begin work on the bug.

The bug must:
 - Be focused on solving a user problem
 - Contain data for all fields in the bug template, so the team can pick it up and begin working immediately

Process: If the unexpected results are new use cases that have been identified, but not yet built, new acceptance criteria and test cases should be captured in a new user story and prioritized by the product owner. 

If the Court is not able to reproduce the bug, add the “Unable to reproduce” tag. This will provide visibility into the type of support that may be needed by the Court. In the event that the Court cannot reproduce the bug, the Court will work with Flexion to communicate what type of troubleshooting help may be needed.

## Definition of Done (Updated 4-14-21)
**Product Owner**
 - [ ]  Bug fix has been validated in the Court's test environment

**Engineering**
 - [ ] Automated test scripts have been written
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Interactors should validate entities before calling persistence methods
 - [ ] Code refactored for clarity and to remove any known technical debt
 - [ ] Deployed to the Court's `test` environment if prod-like data is required. Otherwise, deployed to any `experimental` environment for review.

**Comment by Mwindo** [comment]
The reason this is happening:

- Across the relevant tabs, we validate when components lose focus (e.g., [here](https://github.com/ustaxcourt/ef-cms/blob/263df266f4d323a93eca085d9d03a85de1177272/web-client/src/views/StartCaseInternal/Parties.tsx#L95)) or change selections (e.g., [here](https://github.com/ustaxcourt/ef-cms/blob/263df266f4d323a93eca085d9d03a85de1177272/web-client/src/views/StartCaseInternal/Parties.tsx#L43)) 
- Every time these blur (= unfocus) and selection events happen, we update the error notification but [also automatically scroll to it](https://github.com/ustaxcourt/ef-cms/blob/263df266f4d323a93eca085d9d03a85de1177272/web-client/src/views/ErrorNotification.tsx#L30-L35).
- However, we [ignore validation](https://github.com/ustaxcourt/ef-cms/blob/263df266f4d323a93eca085d9d03a85de1177272/web-client/src/presenter/sequences/validatePetitionFromPaperSequence.ts#L12) when the user hasn't attempted to submit, so no screen jumping occurs at first.

There is a `preventAutoScroll` attribute ... but it can't possibly be being used effectively given that [it is always false](https://github.com/ustaxcourt/ef-cms/blob/263df266f4d323a93eca085d9d03a85de1177272/web-client/src/presenter/computeds/alertHelper.ts#L18).

**Comment by Mwindo** [comment]
In light of conversation with UX, this ticket has evolved into something like, "Across our forms, get the ErrorNotification errors to update live (along with the inline validation errors) and prevent jumping behavior except on submit."

**Comment by Mwindo** [comment]
As I am going through to update forms, I am seeing and fixing a lot of inconsistencies/bugs:

1. Many forms don't live validate properly, either because they are missing live validation (e.g., Status Report Order) or because the validation is misconfigured/not fully configured (e.g., editing some contact information)
2. Some forms allow a 400 rather than validating (on petitions payment information, blank values can trigger this)

**Comment by Mwindo** [comment]
I will push to an experimental branch as soon as one is free so we can compare new behavior with existing behavior on test.

**Comment by Mwindo** [comment]
In the meantime, I will also post a couple of weird behaviors that I discovered (but did not fix), in case we want to address them at some point:

1. When a docket clerk edits a docket record, there can be a mismatch between the validation message and the displayed fields


[Private Zenhub Video](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOWI1Qmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f26b87e8584b662af1a133cccf92dc56e29e20e0/Screen%20Recording%202024-09-20%20at%208.51.45%E2%80%AFAM.mov)

2. In general, validation in the modal can interfere with validation in the background screen, and vice-versa. Here is what occurs in the following video:

- We try to submit a new Trial Session, which says, "Nope--show the validation errors!"
- When we open the modal, that statement is left over and now applies to the modal, too, and so validation errors immediately display on the modal form.
- We correct the validation errors in the modal, which says, "Great--no more validation errors! Clear 'em!" This clears the unrelated validation errors displayed in the background on the Add Trial Session form, too.


[Private Zenhub Video](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBOWo1Qmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--a971b99108a592f99d862b394f92988864e20034/Screen%20Recording%202024-09-24%20at%208.21.11%E2%80%AFAM.mov)

**Comment by Mwindo** [comment]
Another quirk: the docket clerk can deselect a trial location for a docket entry. Is this a bug, or should it be allowed?


[Private Zenhub Video](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBdzc2Qmc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--8e3559379bea78e7ee189989c3e9106bf68c4578/Screen%20Recording%202024-09-24%20at%209.12.23%E2%80%AFAM.mov)

**Comment by mwestereng1** [comment]
UX notes have been added to this [10074 validation error bug](https://docs.google.com/document/d/1_xrpqTiXWroOOF_gr7JoHGOFdGfb1YTBDdXUYyX4mQM/edit?usp=sharing) doc

**Comment by mwestereng1** [comment]
**UX Feedback:**

- [x] When adding a new practitioner, if there is a duplicate email address, the form does not scroll to the top of the page after validating to show the user the error message. 
     ![Image](https://github.com/user-attachments/assets/74542ba0-116d-4dd2-86b1-fdedbb95c341)

- [x] When adding a new practitioner, there is no error message displayed in the banner for mailing address, city, state, or zip code. The inline error message displays correctly. 
     ![Image](https://github.com/user-attachments/assets/bc104d1f-882c-4b30-8c67-212712ba0039)

     ![Image](https://github.com/user-attachments/assets/06a13aa4-2395-49f2-954b-4a54127d948d)

- [x] When an internal court user starts a new petition, if "Deficiency" is selected for "Type of case" and they change the radio dial for from "No" to "Yes" for "IRS Notice provided?" after validating, the Statistic Proposed By IRS does not validate until the form is re-validated.  
     ![Image](https://github.com/user-attachments/assets/42b67c94-5611-4d1a-8009-dd2f203ad9fa)

**Comment by mwestereng1** [comment]
Add Practitioner - Removing fields does not trigger validation (this has been fixed for editing a practitioner)
![Image](https://github.com/user-attachments/assets/e5b319f6-5104-40da-97a3-9edf35931cb2)
![Image](https://github.com/user-attachments/assets/7df196c7-8e71-4e49-8255-ede07c63e029)

**Comment by mwestereng1** [comment]
After a conversation with Christopher, we're not addressing the Add Practitioner - Removing fields does not trigger validation in this story. This behavior already exists in the system. This is also an edge case. It would be best handled in a separate design debt ticket to prevent blocking 10466.

**Comment by Mwindo** [comment]
@ttlenard: Here is a list of substantive changes, which basically amount to "fix the bug, and in the process make validation more consistent across the app." The goal of the work became to be at least as good everywhere, and mostly better. It isn't meant to make things perfectly consistent and to catch everything, but to make them more consistent, which improves the app and makes a later refactor easier. (This task naturally evolved because the fix for the original bug meant updating other forms. Since UX ended up testing almost every form as it was, we figured we'd make easy fixes along the way and note existing issues.)

- Fixes the bug itself
- Fixes a bug in which a validation error was not caught on submission of edited case details (with missing payment info) and instead led to an app error
- Anywhere we use the `ErrorNotification` banner, we keep it in sync with the inline validation. (There are a few existing bugs in which the `ErrorNotification` banner lists fields that are not actually available in the form. We did not address this.)
- Fixes certain issues with the Add/Edit Practitioner forms. This included missing "firstName" and "lastName" having nonstandard validation messages (like `"firstName" is required`); the "this email already exists" error not scrolling to the error to show the user something went wrong (instead, it just looked like the button didn't work); contact fields not showing validation errors (for add practitioner)
- Fixes a duplicate inline validation message ("Enter other filing party name") in Add Paper Filing
- Validation messages on the Status Report Order form and Custom Case Report form now appear and disappear appropriately
- On QCing a document, validation messages now appear and disappear appropriately
- Forms that showed an `ErrorNotification` banner on submit but did not scroll to it now scroll to it. (The Advanced Search Forms, for instance.)
